### PR TITLE
Handle eperm errors windows

### PR DESCRIPTION
**Pull request by wvankuipers**
**This is a fork based on the work of @codeimpossible as the base of PR #13 and with the comments as given by @iarna on that same PR.**

I've implemented a MD5 hash check to verify that the destination file is indeed the same file as the tmp file on Windows when the EPERM error is occurring.

Please note that this is my first PR on a project of this type, please let me know ~~if~~ that I did anything wrong or ~~if~~ that I need to improve on some level.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -90,8 +93,42 @@ function handleClose (writeStream) {
...
+    ) {
+      if (fs.existsSync(writeStream.__atomicTarget)) {
+        fs.readFile(writeStream.__atomicTmp,
+          function (readErr, buf) {
+            if (readErr) return cleanup(err)
+
+            var atomicTmpMD5 = md5(buf)
+            fs.readFile(writeStream.__atomicTarget,
+              function (readErr, buf) {
+                if (readErr) return cleanup(err)
```
Comment (codeimpossible) [inline index.js]: maybe rename `readErr` and above to match the read operation? `readTempErr` and `readTargetErr`?

Comment (wvankuipers) [inline index.js]: Changed it, thanks for the notice!

**Comment by iarna** [comment]
We avoid using md5 because some folks regulatory compliance checks will fail if md5 is used _anywhere_ even in places where the cryptographic properties of the hash don't matter. (They use software to verify this and don't like exceptions.)

Could you do this using something like `sha256` from [`crypto`](https://nodejs.org/api/crypto.html#crypto_class_hash)?

**Comment by wvankuipers** [comment]
I integrated Crypto in favor of MD5, also fixed an issue with the tmp file not being removed if the copy failed and the target file is the same as the tmp file.

**Comment by turp** [comment]
Any idea when this PR will get merged in? A lot of windows developers are running into this issue.

Thanks!

File: index.js [hunk 2]
```diff
@@ -90,8 +93,49 @@ function handleClose (writeStream) {
       writeStream.emit('close')
     })
   }
+  function trapWindowsEPERMRename (err) {
+    if (err.syscall && err.syscall === 'rename' &&
+        err.code && err.code === 'EPERM' &&
+        writeStream.__isWin
+    ) {
+      if (fs.existsSync(writeStream.__atomicTarget)) {
```
Comment (iarna) [inline index.js]: It strikes me as weird that this is a sync call and the `readFile` calls are async. Probably these should all be async.

File: index.js [hunk 3]
```diff
@@ -90,8 +93,49 @@ function handleClose (writeStream) {
...
+        var tmpHash = crypto.createHash('sha256')
+        var targetHash = crypto.createHash('sha256')
+
+        fs.readFile(writeStream.__atomicTmp,
```
Comment (iarna) [inline index.js]: I would also prefer to not have a cascade of callbacks here and instead use named functions for each layer of callbacks. Keeps the code flatter and the callback names have a self-documenting property.

File: index.js [hunk 4]
```diff
@@ -90,8 +93,49 @@ function handleClose (writeStream) {
...
+          function (readTmpErr, tmpBuffer) {
+            if (readTmpErr) return cleanup(err)
+
+            tmpHash.update(tmpBuffer)
+            fs.readFile(writeStream.__atomicTarget,
+              function (readTargetErr, targeBuffer) {
```
Comment (iarna) [inline index.js]: s/targeBuffer/targetBuffer/ ?

File: index.js (hunk 3 above)
Comment (iarna) [inline index.js]: The only slightly jinky bit here is that on error you want to call `cleanup()` with the top-level error, so just passing `cleanup` to `iferr` wouldn't work. But all that calls for is wrapping `cleanup` with something that does that.

**Comment by codeimpossible** [comment]
@iarna thanks for the feedback on the code, i don't think i can update this PR so I'm happy if you're cool with you fixing it ;)

I'll close out the other PR too since it's not needed anymore.

**Comment by iarna** [comment]
@codeimpossible Oh yeah, good point =D

File: index.js (hunk 3 above)
Comment (wvankuipers) [inline index.js]: Hi @iarna, thank you for the feedback. I made some changes to fix the issues you pointed out, but I'm not quite sure about this one

> The only slightly jinky bit here is that on error you want to call cleanup() with the top-level error, so just passing cleanup to iferr wouldn't work. But all that calls for is wrapping cleanup with something that does that.

Could you elaborate on this so I can implement this as well?

**Comment by Singleton06** [comment]
Out of curiosity, what's the status of this pull request?  We've been running into quite a few issues with this in our development.

**Comment by mateodelnorte** [comment]
Is this going to be merged?

**Comment by mateodelnorte** [comment]
@wvankuipers just a note - I've been investigating this as a fix for https://github.com/npm/npm/issues/12059#issuecomment-244663459, as stated in the comments there. Using this branch of fs-write-stream-atomic and a current version of npm (4.1.1) does not solve my `EPERM: operation not permitted, rename` problem in Windows 7.

**Comment by wvankuipers** [comment]
@mateodelnorte The issue you are referring to has multiple different issues combined. some of these issues will be corrected by this PR, but others are from package that don't use fs-write-stream-atomic so they will not be fixed. My guess is that other fs packages have the same issue, those should be fixed separately.

**Comment by domenkozar** [comment]
How can we get this merged? Are there any workaround using this PR? Is building our own node the only way?

**Comment by wvankuipers** [comment]
@domenkozar No clue about if this is going to be merged, there seems to be a radio silence from @iarna regarding this issue 😞

**Comment by ivankolev** [comment]
I've had success updating to npm > 4, which made those errors disappear. I
think I saw something in the changelogs about that.

**Comment by domenkozar** [comment]
Thanks for the tip!

**Comment by iarna** [comment]
Ok, so here's what I've done:

I rewrote it a bit to read the file hashes using streams (as we don't know how big these files are, possibly tens or hundreds of megabytes).

I switched to sha-512 (which is faster on 64 bit systems).

I squashed all the commits down to one.

This is now merged as 2f51136.

**Pull request closed without merging**